pkg/db/db.go

```diff
@@ -0,0 +1,118 @@
+package db
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/red-hat-storage/managed-fusion-fleet-reconciler/pkg/types"
+	"go.uber.org/zap"
+)
+
+func GetConnectionString(host, user, password, name string, port int) string {
+	return fmt.Sprintf(
+		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
+		host,
+		port,
+		user,
+		password,
+		name,
+	)
+}
+
+type Database struct {
+	tables struct {
+		providers string
+	}
+	pool *pgxpool.Pool
+	// conn is used to listen for notifications and should be closed
+	conn *pgxpool.Conn
+}
+
+func NewClient(ctx context.Context, connString string, tables map[string]string) (*Database, error) {
+	pool, err := pgxpool.New(ctx, connString)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new instace of pool: %v", err)
+	}
+
+	if err := pool.Ping(ctx); err != nil {
+		return nil, fmt.Errorf("failed to ping database: %v", err)
+	}
+
+	if _, ok := tables["providers"]; !ok {
+		return nil, fmt.Errorf("incomplete table name mapping, missing a table name for \"providers\" table")
+	}
+
+	db := Database{}
+	db.pool = pool
+	db.tables.providers = tables["providers"]
+	return &db, nil
+}
+
+func (pg *Database) Close(ctx context.Context) {
+	if pg.conn != nil {
+		pg.conn.Conn().Close(ctx)
+	}
+	pg.pool.Close()
+}
+
+func (pg *Database) OnProvider(ctx context.Context, logger *zap.Logger, notifyExisting bool, fn func(providerName string)) error {
+	var err error
+	if pg.conn, err = pg.pool.Acquire(ctx); err != nil {
+		return fmt.Errorf("failed to acquire connection: %v", err)
+	}
+
+	query := fmt.Sprintf("LISTEN %s;", pg.tables.providers)
+	// Listen for notifications
+	if _, err = pg.conn.Exec(ctx, query); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			notification, err := pg.conn.Conn().WaitForNotification(ctx)
+			if err != nil {
+				// If the connection is closed, release the connection and exit the goroutine
+				if pg.conn.Conn().IsClosed() {
+					pg.conn.Release()
+					logger.Error("Connection closed by server:", zap.Error(err))
+					return
+				}
+				logger.Error("failed to wait for notification", zap.Error(err))
+				continue
+			}
+			fn(notification.Payload)
+		}
+	}()
+
+	if notifyExisting {
+		go func() {
+			query := fmt.Sprintf("SELECT cluster_id FROM %s", pg.tables.providers)
+			rows, err := pg.pool.Query(ctx, query)
+			if err != nil {
+				logger.Fatal("failed to query database", zap.Error(err))
+			}
+			defer rows.Close()
+
+			for rows.Next() {
+				var clusterID string
+				if err := rows.Scan(&clusterID); err != nil {
+					logger.Fatal("failed to scan row", zap.Error(err))
+				}
+				fn(clusterID)
+			}
+		}()
+	}
```

Comment:
I am not sure this addresses the concern raised here: https://github.com/red-hat-storage/managed-fusion-fleet-reconciler/pull/5#discussion_r1266730976

Mainly as you are just invoking 2 go routines one after another, but I am not sure go has any kind of guarantees that the second go routine will start execution after the first one. This might result in very subtle bugs.

Reply:
Yes, we are invoking 2 go routines one after the other after subscribing for the notification it doesn't matter which go routine runs first, one goroutine read from the notification and another reads from the database and processes the data. The point to note here is that we are not subscribing to the events in the goroutine it's done before starting the goroutines,  AFAIK it addresses the concern raised above. Let me know if it makes sense or if we need to change anything here.

Reply:
Ok, I understand now. Thanks for the clarification!